In [3]:
%load_ext sql
%sql mysql+pymysql://root:@localhost:3307/win

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


<h1 style="text-align:center; color:#005bbd; font-size:50px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">RANGE Mode</h1>


<br><br>
<div style="text-align:justify; color:black; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;"><p>The RANGE mode is a clause used to define the window frame for calculations based on the values in the ORDER BY clause. The RANGE mode differs from the ROWS mode in that it considers the values of the ordering column(s) rather than a fixed number of rows.<br><br><ul>In RANGE mode, the window frame is not determined by a specific number of rows but is based on the actual values in the ordering column(s).<br><br>
The window frame includes all rows with values equal to the current row's values according to the ORDER BY clause.<br><br>
    It allows for a more flexible definition of the window based on the values rather than a fixed number of rows.</ul><br>
</p>
</div><br>







<br><br>
<div style="text-align:justify; color:red; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;">A general (and much simplified) format of a window function call is: <ul>function_name OVER (PARTITION BY ... ORDER BY ... frame_clause)</ul></div><br>








<div style="text-align:justify; color:blue; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;">frame_clause : <ul>mode BETWEEN frame_start AND frame_end [ frame_exclusion ]</ul><br>here i use Range mode SO : <br>Syntax is <br><br>RANGE BETWEEN frame_start AND frame_end [ frame_exclusion ]</div><br>


In [4]:
%%sql 
SELECT * FROM tmp


 * mysql+pymysql://root:***@localhost:3307/win
5 rows affected.


[(5.5,), (6.0,), (6.0,), (7.0,), (6.0,)]

<h1 style="text-align:center; color:blue; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">ROWS/RANGE BETWEEN 1 PRECEDING AND CURRENT ROW </h1>


In [5]:
%%sql 
SELECT  val, 

SUM(val) OVER (ORDER BY val DESC ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) AS row_mode,
SUM(val) OVER (ORDER BY val DESC RANGE BETWEEN 1 PRECEDING AND CURRENT ROW) AS range_mode 

FROM tmp;

 * mysql+pymysql://root:***@localhost:3307/win
5 rows affected.


[(7.0, 7.0, 7.0),
 (6.0, 13.0, 25.0),
 (6.0, 12.0, 25.0),
 (6.0, 12.0, 25.0),
 (5.5, 11.5, 23.5)]

<div style="text-align:justify; color:black; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;"> Lets Understand Hows range mode are different</div><br><img style="display: block;margin-left: auto;margin-right: auto;width: 80%; " src="./sql_images/range.png"><br><br>
<br>


<h1 style="text-align:center; color:blue; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">ROWS/RANGE BETWEEN 1 PRECEDING AND 1 FOLLOWING </h1>


In [7]:
%%sql 
SELECT  val, 

SUM(val) OVER (ORDER BY val DESC ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING ) AS row_mode,
SUM(val) OVER (ORDER BY val DESC RANGE BETWEEN 1 PRECEDING AND 1 FOLLOWING) AS range_mode 

FROM tmp;

 * mysql+pymysql://root:***@localhost:3307/win
5 rows affected.


[(7.0, 13.0, 25.0),
 (6.0, 19.0, 30.5),
 (6.0, 18.0, 30.5),
 (6.0, 17.5, 30.5),
 (5.5, 11.5, 23.5)]

<div style="text-align:justify; color:black; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;"> Lets Understand</div><br><img style="display: block;margin-left: auto;margin-right: auto;width: 80%; " src="./sql_images/range_2.png"><br><br>
<br>


<h1 style="text-align:center; color:blue; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">ROWS/RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING </h1>


In [20]:
%%sql 
SELECT  val, 

SUM(val) OVER (ORDER BY val DESC ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING ) AS row_mode,
SUM(val) OVER (ORDER BY val DESC RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS range_mode 

FROM tmp;

 * mysql+pymysql://root:***@localhost:3307/win
5 rows affected.


[(7.0, 30.5, 30.5),
 (6.0, 30.5, 30.5),
 (6.0, 30.5, 30.5),
 (6.0, 30.5, 30.5),
 (5.5, 30.5, 30.5)]

<br>
<div style="text-align:justify; color:black; font-size:20px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;"> Here UNBOUNDED PRECEDING is 7.0 and UNBOUNDED FOLLOWING is 5.5 for both Rows and Range mode<br><ul>SO The Sum of (7.0) , 6.0,6.0,6.0 , (5.5) is 30.5</ul></div><br>
<br>


<h1 style="text-align:center; color:blue; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">ROWS/RANGE BETWEEN UNBOUNDING PRECEDING AND  CURRENT ROW</h1>


In [29]:
%%sql 
SELECT  val, 

SUM(val) OVER (ORDER BY val DESC ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW ) AS row_mode,
SUM(val) OVER (ORDER BY val DESC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS range_mode 

FROM tmp;

 * mysql+pymysql://root:***@localhost:3307/win
5 rows affected.


[(7.0, 7.0, 7.0),
 (6.0, 13.0, 25.0),
 (6.0, 19.0, 25.0),
 (6.0, 25.0, 25.0),
 (5.5, 30.5, 30.5)]

<div style="text-align:justify; color:black; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;"> Lets Understand</div><br><img style="display: block;margin-left: auto;margin-right: auto;width: 80%; " src="./sql_images/range_3.png"><br><br>
<br>


<div style="text-align:justify; color:black; font-size:25px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;"> We also use Partition for perform operations in groupwise</div>


In [45]:
%%sql 
SELECT  val, 

SUM(val) OVER (PARTITION BY val ORDER BY val DESC ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW ) AS row_mode,
SUM(val) OVER (PARTITION BY val ORDER BY val DESC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS range_mode 

FROM tmp;

 * mysql+pymysql://root:***@localhost:3307/win
5 rows affected.


[(5.5, 5.5, 5.5),
 (6.0, 6.0, 18.0),
 (6.0, 12.0, 18.0),
 (6.0, 18.0, 18.0),
 (7.0, 7.0, 7.0)]

In [75]:
%%sql 
SELECT  val, 

SUM(val) OVER total AS 'sum'

FROM tmp

WINDOW total AS (
PARTITION BY val 
ORDER BY val DESC 
ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING 
)


 * mysql+pymysql://root:***@localhost:3307/win
5 rows affected.


[(5.5, 5.5), (6.0, 18.0), (6.0, 18.0), (6.0, 18.0), (7.0, 7.0)]

<div style="text-align:left; color:black; font-size:20px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;">Summary
    <ul>What is Range</ul>
    <ul>ROWS, GROUPS(Not supported in Mysql), RANGE Modes</ul>
    <ul>UNBONDED PRECEDING  (start row)</ul>
    <ul>UNBONDED FOLLOWING  (end row)</ul>
    <ul>X PRECEDING  (start row no with X)</ul>
    <ul>X FOLLOWING  (end row no with X)</ul>
    <ul>CURRENT ROW  (end current row selected)</ul>
    <ul>Also we can use PARTITION BY</ul>
    <ul>Alias WINDOW keyword</ul>


</div>
<hr>
<div style="text-align:left; color:black; font-size:20px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black;">Commands
<i style="color:green;">
<ul>ORDER BY col DESC/ASC RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING</ul>
    <ul>PARTITION BY col1 ORDER BY col2 DESC/ASC RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING</ul>
        <hr>
<ul>ORDER BY col DESC/ASC RANGE BETWEEN X PRECEDING AND X FOLLOWING</ul>
<ul>PARTITION BY col1 ORDER BY col2 DESC/ASC RANGE BETWEEN X PRECEDING AND X FOLLOWING</ul>
    <hr>
<ul>ORDER BY col DESC/ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW</ul>
<ul>PARTITION BY col1 ORDER BY col2 DESC/ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW</ul>

    
</i>

</div>